SSO redirect

mattporritt opened the pull request:
Add functionality via plugin config to stop users from accessing the Moodle login page and instead redirect them to the SSO "host". The URL to redirect to can be set in the Plugin config.
Also add the option for users to override this redirect and to access the regular Moodle Login page by useing the query parameter: ?enrolkey_skipsso=on

dmitriim commented on auth.php:
```diff
@@ -69,6 +71,50 @@ public function __construct() {
     }
 
     /**
+     * All the checking happens before the login page in this hook
```
Can we please add unit tests for pre_loginpage_hook() and loginpage_hook()

dmitriim commented on auth.php:
```diff
@@ -69,6 +71,50 @@ public function __construct() {
+        // If SSO only is set and user is not passing the skip param
+        // or has it already set in their session then redirect to the SSO URL.
+        if ($this->config->ssoonly == 1 && !$skipsso) {
+            redirect($this->config->ssourl);
```
What's going to happen if ssourl is empty? We need to add this check when we save config. If we set ssoonly = yes, then we have to add ssourl.

brendanheywood commented:
@mattporritt can I suggest you rename the param to something which involves 'enrolkey' just so there is zero chance this clashes with other auth plugins. Also please rebase onto latest master

brendanheywood commented on auth.php:
```diff
@@ -69,6 +70,50 @@ public function __construct() {
+        // or has it already set in their session then redirect to the SSO URL.
+        if ($this->config->ssourl != '' && !$skipsso) {
+            redirect($this->config->ssourl);
+        } else {
```
the double redundant return is a little weird

mattporritt commented:
I've removed some of the redundant logic. I've also refactored the code out into the should_redirect() pattern, to make it easier to test.
There are tests covering the use cases now. I didn't add an explicit test for pre_loginpage_hook() as there really isn't any point. The only thing testing that method would do is test one core PHP function. Also Moodle requires that method to have it redirect to another method instead of returning a value.

brendanheywood commented on auth.php:
```diff
@@ -253,7 +278,7 @@ protected function should_create_user() {
 
     /**
-     * Check if restriction by IP is enabled.
+     * Check if restriction by IP        var_dump(print_r($this->auth,true));is enabled.
```
please remove this

mattporritt replied:
I have no idea how that even got there... but will get rid of it

brendanheywood commented on lang/en/auth_userkey.php:
```diff
@@ -38,4 +38,7 @@
+$string['pluginisdisabled'] = 'The userkey authentication plugin is disabled.';
+$string['ssourl'] = 'URL of SSO host';
+$string['ssourl_desc'] = 'URL of the SSO host to redirect users to. '.
+                         'If defined users will be redirected here on login instead of the Moodle Login page';
```
there is a bug in AMOS where it can't correctly parse string like this, please make this a single line. Pretty sure the travis linter is aware of this and won't complain inside lang files

mattporritt replied:
just us humans who can't read it

brendanheywood commented on tests/auth_plugin_test.php:
```diff
@@ -708,4 +709,33 @@ public function test_that_return_wantsurl_if_it_is_external_url() {
+        set_config('ssourl', 'http://google.com', 'auth_userkey');
+        $this->auth = new auth_plugin_userkey();
+        $userredirect = $this->auth->should_login_redirect();
+        $this->assertEquals($userredirect, false);
```
we should probably test all the negative cases here too

dmitriim replied:
Actually I would rather get rid of these tests and test actual hook as hook is getting called and we can change the code of the hook in some point and break things. 
Look at https://github.com/CatalystIT-AU/moodle-auth_userkey/blob/master/tests/auth_plugin_test.php#L111 as an example 

should_login_redirect() can be protected as it's not really important function outside of the class. Having it in the code is more about readability improvements rather then functionality.

brendanheywood replied:
This is by design, you can't test redirection in unit tests, only behat. It's probably not worth effort to swap

dmitriim replied:
As an option you can wrap redirection in a function and test that this function gets called in a hook method with correct data. We need tests for public hook function. Otherwise tests are sort of useless.

dmitriim replied:
The other option maybe catching the exception moodle_exception('redirecterrordetected', 'error') in your unit tests. 
But unfortunately it doesn't say us what URL we are going to redirect a user.

dmitriim commented on auth.php:
```diff
@@ -413,11 +438,38 @@ public function get_request_login_url_user_parameters() {
+     *
+     * @return bool
+     */
+    public function should_login_redirect() {
```
This function can be protected as we will never use it outside of the class.

dmitriim commented:
Will be merged as part of https://github.com/CatalystIT-AU/moodle-auth_userkey/pull/8
